fix --compile=all for jb/functions

vtjnash opened the pull request:
(note: this is against the jb/functions branch PR)

this simplifies the implementation of sparams (by always keeping them with the LambdaInfo and adding a calling convention such that they can be passed directly) and then unspecialized (since there's less need for it to be sparam-aware) and also instantiate_staged (since there's no need for it to edit the ast following the changes to sparams). From one measurement, this appeared to result in a net 0.5% reduction in sysimg size.

this also includes improvements to the `--compile=all` algorithm to be able to handle some changes to Base in the past month.

vtjnash committed:
refactor linfo.sparams

Moves the sparam information from the AST to the LambdaInfo so that it
is more accessible to method dispatch (among other uses)

vtjnash committed:
pass sparam_vals through to the jlcall if it wasn't known at compile-time

this marks the low bit of the fptr to indicate to the frontend which jl_fptr_t calling convention should be used

vtjnash committed:
also improve the tvar union splitting algorithm to handle more than one tvar

vtjnash committed:
fix --compile=all

and fix a method definition mistake in base which defined
convert(T, Float64) where T could be a UnionType instead of a leaftype

vtjnash committed:
extend the --compile=all union splitting algorithm to handle unions that appear in the type signature directly

vtjnash committed:
simplify linfo->unspecialized and staged function instantiation

since the sparams env can now be passed directly into a function as an extra argument,
linfo->unspecialized can (usually) be shared among all instances of a function
(the only exception being the presence of intrinsics that are expecting to be able to call static_eval);
the implementation of jl_instantiate_staged can be greatly simplified;
and code_llvm can safely generate code for any function (via unspecialized), without corrupting the cached AST info

vtjnash committed:
make spenv handling in inlining a bit more efficient

JeffBezanson commented on src/julia_internal.h:
```diff
@@ -55,11 +55,23 @@ void jl_compile_linfo(jl_lambda_info_t *li, void *cyclectx);
+        return ((jl_fptr_sparam_t)(fptr_int & ~mask))(meth->sparam_vals, args[0], &args[1], nargs-1);
+
+
+    if (__likely((((uintptr_t)fptr) & 1) == 0))
```
?

vtjnash replied:
needs comments -- noted.

JeffBezanson commented:
Haven't fully reviewed this yet but + :100: .

JeffBezanson commented on src/gf.c:
```diff
@@ -278,30 +278,41 @@ static jl_lambda_info_t *jl_method_table_assoc_exact(jl_methtable_t *mt, jl_valu
 {
     JL_GC_PUSH1(&sp);
-    if (jl_svec_len(l->sparams) > 0)
-        sp = jl_svec_append(sp, l->sparams);
```
I think `jl_svec_append` is unused now and can be removed.

vtjnash committed:
use explicit jlcall_api flag instead of stealing bits from the fptr

x86 doesn't guarantee function-pointer alignment, and adding the gcc attribute to force it is probably not portable
it shouldn' be particularly critical from a performace perspective anyhow

vtjnash commented on src/gf.c:
```diff
@@ -2032,7 +2075,9 @@ void jl_add_method_to_table(jl_methtable_t *mt, jl_tupletype_t *types, jl_lambda
         // already used by another GF; make a copy (issue #10373)
+        assert(meth->sparam_vals == jl_emptysvec);
         meth = jl_add_static_parameters(meth, jl_emptysvec, NULL);
+        meth->unspecialized = NULL;
```
this code will need a new test (the existing test from #10373 no longer invokes it because of the special toplevel closure handling) has already caused the lambda to be newly created..

for example, the following:
```julia
for f in (cos, sin, tan)
    (ft::typeof(f))(x::newtype) = println(ft)
end
```
